heron/schedulers/src/java/com/twitter/heron/scheduler/mesos/framework/MesosFramework.java

```diff
@@ -0,0 +1,541 @@
+    // TODO(mfu): TO handle this failure
+    LOG.info("Slave lost: " + slaveID.toString());
+  }
+
+  @Override
+  public void executorLost(SchedulerDriver schedulerDriver,
+                           Protos.ExecutorID executorID,
+                           Protos.SlaveID slaveID,
+                           int status) {
+    // TODO(mfu): TO handle this failure
+    LOG.info(String.format("Executor lost: %s, slave ID: %s, status: %d",
+        executorID.toString(), slaveID.toString(), status));
+  }
+
+  @Override
+  public void error(SchedulerDriver schedulerDriver, String message) {
+    // TODO(mfu): TO handle this failure
+    LOG.info("Received error: " + message);
+  }
+
+
+  /**
+   * Generate launchable tasks basing on offer resources
+   *
+   * @param offerResources Mapping from Protos.Offer to utils class TaskResources
+   * @return list of launchable task
+   */
+  protected List<LaunchableTask> generateLaunchableTasks(
+      Map<Protos.Offer, TaskResources> offerResources) {
+    List<LaunchableTask> tasks = new LinkedList<>();
+
+    if (isTerminated) {
+      LOG.info("Job has been killed");
+      return tasks;
+    }
+
+    while (!toScheduleTasks.isEmpty()) {
+      String taskId = toScheduleTasks.poll();
+      BaseContainer baseContainer =
+          containersInfo.get(TaskUtils.getContainerIndexForTaskId(taskId));
+
+      TaskResources neededResources =
+          new TaskResources(
+              baseContainer.cpu,
+              baseContainer.memInMB,
+              baseContainer.diskInMB,
+              baseContainer.ports);
+
+      Iterator<Map.Entry<Protos.Offer, TaskResources>> it = offerResources.entrySet()
+          .iterator();
+      while (it.hasNext()) {
+        Map.Entry<Protos.Offer, TaskResources> kv = it.next();
+        Protos.Offer offer = kv.getKey();
+        TaskResources resources = kv.getValue();
+
+        if (resources.canSatisfy(neededResources)) {
+          resources.consume(neededResources);
+
+          // Add the tasks
+
+          // Construct the list of free ports to use for a heron executor.
+          List<Integer> freePorts = new ArrayList<>();
+          for (int port = (int) (neededResources.getPortsHold().get(0).rangeStart);
+               port <= (int) (neededResources.getPortsHold().get(0).rangeEnd); port++) {
+            freePorts.add(port);
+          }
+
+          tasks.add(new LaunchableTask(taskId, baseContainer, offer, freePorts));
+
+          // Matched baseContainer, break;
+          break;
+        } else {
+          LOG.info(String.format("Insufficient resources remaining for baseContainer: %s, "
+                  + "will append to queue. Need: [%s], Found: [%s]",
+              taskId, neededResources.toString(), resources.toString()));
+
+          toScheduleTasks.add(taskId);
+
+          // No enough Resources remaining. Return the tasks generated.
+          return tasks;
```

Comment:
The "else" case means the current resource offer cannot meet the request. But we can try remaining resource offers  using iterator instead of returning it directly.